docs/enhancements/using-dtk-for-building-driver-containers.md

```diff
@@ -0,0 +1,127 @@
+      generation: 2
+      image: sha256:84edddb4322cfe50d8545c41014a8f5595a4260a20d9a269e144909c5949582b
+    tag: latest
+```
+
+We can see that:
+* We have a different tag for each RHCOS version that the cluster has known.
+* The ‘latest’ tag point to the most recent RHCOS tag in the imagestream
+* The `status` field of the imagestream keeps an array of images for each tag with the full digest history used for that tag
+    * The latest tag was first pointing to `410.84.202207140725-0` tag and then move to point to `411.86.202207260413-0` tag.
+
+
+## Node reconciler flow
+Create and update an in-memory mapping `kernelVersion` → `rhcosVersion`.
+* The `kernelVersion` can be found on the node object at `node.status.nodeInfo.kernelVersion`
+* The `rhcosVersion` can be found on the node object at `node.status.nodeInfo.osImage`
+
+## Module reconciler flow
+
+When a build pod is created, or when a cluster upgrade kicks, KMMO will get the list of kernels running in the cluster from the node labels.
+For each kernel in the cluster, if a build is required, get the matching `rhcosVersion` from the `kernelVersion` → `rhcosVersion` mapping. If there is more than one we can just pick the latest RHCOS available version with the required kernel.
+KMMO will use that RHCOS version as the imagestream tag in order to build/rebuild the driver container.
+
+## Dockerfile examples
+If the user want us to pull the correct DTK image on builds/cluster-upgrade, he will need to put a placeholder in the Dockerfile for KMMO to substitute. In case this placeholder isn't present in the Dockerfile, KMMO will pull the base image specified by the user.
+
+#### Dockerfile which let KMMO find the correct DTK image.
+```
+FROM DTK-AUTO
+RUN gcc ...
+...
+```
+
+#### Dockerfile with 0-magic.
+```
+FROM user.registry.com/org/user-base-image:tag
+RUN gcc ...
+...
+```
+
+## Disconnected clusters
+This approach is expected to work also in disconnected clusters as we don’t need to pull the driver-toolkit image manually to get the kernel RPMs installed in it. Only the buildconfig will pull the DTK image, therefore, it will be redirected if the user has set an `ImageSourceContentPolicy` in the cluster.
+
+## Pre-built driver containers
+
+This case is simpler because unlike KMMO, which is surprised by a cluster upgrade when he sees a node with a new kernel label, a user upgrading his cluster should know to what OCP version he is upgrading to, therefore, it can get the DTK image by running
+
+`oc adm release info quay.io/openshift-release-dev/ocp-release:<cluster-version>-x86_64 --image-for=driver-toolkit` and getting the relevant driver-toolkit image for is pre-built driver-container.
+
+## Future possible usage
+KMMO can watch the image stream and act upon image changes. For example KMMO can rebuild a driver-container with a newer version (for the same z-stream) available in the imagestream.
```

Comment:
I think we could only do that if we parsed the Dockerfile and look at the `FROM` directives.

Reply:
i am not sure that this is something we need to consider right now. I think KMMO should build image only on specific request (preflight/prebuild) or when the image is missing during deployment.Otherwise we might interfere with user-built images

Reply:
Indeed, this is only possible if the user as set the `DRIVER-TOOLKIT` place holder in his Dockerfile.

Reply:
Resolving the thread.